pop_tools/budget.py

```diff
@@ -0,0 +1,399 @@
+    ds = _subset_slice_mask(ds, var_list=['WT'], kmax=kmax + 1, mask=mask)
+    ds = _convert_to_tendency(ds, grid, kmax=kmax + 1)
+
+    # Compute divergence of vertical advection.
+    # NOTE: Is there a case where WT_{tracer} would be saved out as a vertical
+    # integral? In that case, this would break.
+    vadv = (ds.WT.shift(z_t=-1).fillna(0) - ds.WT).isel(z_t=slice(0, -1))
+    vadv = vadv.rename('vadv')
+    vadv = _convert_units(vadv)
+    vadv.attrs['long_name'] = 'vertical advection'
+    return vadv
+
+
+def _compute_vertical_mixing(ds, grid, mask, kmax=None):
+    """Compute contribution from vertical mixing."""
+    ds = _subset_slice_mask(ds, var_list=['DIA_IMPVF', 'KPP_SRC'], kmax=kmax, mask=mask)
+
+    # Only need to flip sign of DIA_IMPVF.
+    ds['DIA_IMPVF'] = _convert_to_tendency(ds['DIA_IMPVF'], grid, sign=-1, force_area=True)
+    ds['KPP_SRC'] = _convert_to_tendency(ds['KPP_SRC'], grid, kmax=kmax)
+
+    # Compute divergence of diapycnal mixing
+    diadiff = _compute_vertical_divergence(ds.DIA_IMPVF)
+
+    # Sum KPP and diapycnal mixing to create vmix term.
+    vmix = (ds.KPP_SRC + diadiff).rename('vmix')
+    vmix = _convert_units(vmix)
+    vmix.attrs['long_name'] = 'vertical mixing'
+    return vmix
+
+
+def _compute_SMS(ds, grid, mask, kmax=None):
+    """Compute SMS term from biology."""
+    ds = _subset_slice_mask(ds, var_list=['SMS'], kmax=kmax, mask=mask)
+    ds = _convert_to_tendency(ds, grid, kmax=kmax)
+    ds = _convert_units(ds)
+    ds = ds.rename({'SMS': 'sms'})
+    ds.attrs['long_name'] = 'source/sink'
+    return ds
+
+
+def _compute_surface_flux(ds, grid, mask):
+    """Computes virtual fluxes of tracer."""
+    ds = _subset_slice_mask(ds, var_list=['STF'], mask=mask)
+    ds = _convert_to_tendency(ds, grid)
+    ds = _convert_units(ds)
+    stf = ds.rename({'STF': 'stf'})
+    stf.attrs['long_name'] = 'surface tracer flux'
+    return stf
+
+
+def _compute_virtual_flux(ds, grid, mask):
+    """Computes virtual fluxes of tracer."""
+    ds = _subset_slice_mask(ds, var_list=['FvICE', 'FvPER'], mask=mask)
+    ds = _convert_to_tendency(ds, grid)
+    ds = _convert_units(ds)
+    vf = (ds.FvICE + ds.FvPER).rename('vf')
+    vf.attrs['long_name'] = 'virtual flux'
+    return vf
+
+
+def _process_input_dataset(ds):
+    """Checks that input dataset has appropriate variables, etc."""
+    mandatory_vars = ['UE', 'VN', 'WT', 'HDIFE', 'HDIFN', 'HDIFB', 'DIA_IMPVF', 'KPP_SRC']
+    if not set(mandatory_vars).issubset(ds):
+        missing_vars = [v for v in mandatory_vars if v not in ds]
+        error_msg = (
+            'Input dataset does not contain the mandatory variables for budget analysis. '
+            + f'`ds` is missing {missing_vars}'
+        )
+        raise IOError(error_msg)
+
+    # Drop coordinates for shift, roll, etc.
+    coord_vars = [c for c in ds.coords if c != 'time']
+    ds = ds.drop(coord_vars)
+    if 'z_w_top' in ds.dims:
```

Comment:
Should drop these. Want to return object with proper coordinate labeling. Will handle coordinate alignment by using dask array arithmetic (i.e., by using `.data`).